File: q101-125/q105.py
# 2020.01.11
# Problem Statement:
# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
# Glad to come up with the solution on my own :-)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        # check empty cases
        if len(preorder) == 0: return None
        
        # overall logic:

        # the first node in preorder must be the root

        # if the index of the root in inorder list is x
        # from index 0 to index x-1 (inclusive) would be the nodes of the left sub-tree, update the inorder list for left sub-tree
        # from index x+1 to the last index (inclusive) would be the nodes of the right sub-tree, update the inorder list for right sub-tree

        # connection
        # connect the left of the root to the second node in preorder
        # connect the right of the root to the (1+number of nodes in left sub-tree+1)th node in preorder

        # update preorder list to be [the node after root :]
        # update preorder list to be [the node after (root + all left sub-tree nodes) :]
        
        root = TreeNode(preorder[0])
        self.buildTreeHelper(preorder, inorder, root)
        return root
    
    def buildTreeHelper(self, preorder, inorder, root):
        # base case
        if len(inorder)*len(preorder) == 0: return
        
        if inorder_idx != 0:
            left_inorder = inorder[:inorder_idx]
        else:
            left_inorder = []
        
        if inorder_idx != len(inorder)-1:
            right_inorder = inorder[inorder_idx+1:]
        else:
            right_inorder = []
        
        if len(left_inorder) != 0:
            root.left = TreeNode(preorder[1])
        
        if len(right_inorder) != 0:
            root.right = TreeNode(preorder[len(left_inorder)+1])
        
        self.buildTreeHelper(preorder[1:], left_inorder, root.left)
        self.buildTreeHelper(preorder[len(left_inorder)+1:], right_inorder, root.right)

File: Blind75/q105.java
// 2022.09.25 midnight
// Problem Statement:
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

// idea: for each node, find its index in inorder, and everything btw left bound and index is left tree,
// and everything btw index and right bound goes to the right tree,
// build left subtree and right subtree and merge and return
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/discuss/34538/My-Accepted-Java-Solution
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public HashMap<Integer, Integer> ht;
    public TreeNode buildTreeHelper(int[] preorder, int[] inorder, int p_start, int i_start, int i_end) {
        // base cases to stop
        if (i_start>i_end) return null;
        if (p_start>=preorder.length) return null;
        
        TreeNode new_node = new TreeNode (preorder[p_start]);
        int idx = ht.get(preorder[p_start]);
        // left child is p_start+1 in preorder
        TreeNode left = buildTreeHelper(preorder, inorder, p_start+1, i_start, idx-1);
        // right child's index is 1+left tree's nodes number apart from p_start
        TreeNode right = buildTreeHelper(preorder, inorder, p_start+idx-i_start+1, idx+1, i_end); 
        
        new_node.left = left;
        new_node.right = right;
        return new_node;
        
        
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        ht = new HashMap<> ();
        for (int i=0; i<inorder.length; i++) {
            ht.put(inorder[i], i);
        }
        
        return buildTreeHelper(preorder, inorder, 0, 0, inorder.length-1);
    }
}